### DISCUSSION TITLE
Draw fps capped at 60 with Direct3D 11 even though the limiter is not set to vsync

### DISCUSSION THREAD

**Discussion by mouzedrift** [ppy/osu#24168] [Q&A]
### Type

Performance

### Bug description

When I use Direct3D 11, my fps are capped at 60 fps even when my fps limiter is not set to vsync. This only happens in fullscreen and borderless fullscreen, and the windowed mode doesn't seem to be limiting my fps. The OpenGL and OpenGL (Legacy) renderers don't seem to do that. I can't test Vulkan because my GPU doesn't support it.

### Version

2023.326.0

### Logs

[performance.log](https://github.com/ppy/osu/files/11072364/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11072365/runtime.log)
[updater.log](https://github.com/ppy/osu/files/11072366/updater.log)
[database.log](https://github.com/ppy/osu/files/11072367/database.log)
[input.log](https://github.com/ppy/osu/files/11072368/input.log)
[network.log](https://github.com/ppy/osu/files/11072369/network.log)

**Comment by catuhana** [comment]
Can reproduce, but it happens on windowed mode too for me.

**Comment by Hasster1** [comment]
Try disabling fullscreen optimizations.

**Comment by catuhana** [comment]
> Try disabling fullscreen optimizations.

Still can reproduce.

**Comment by peppy** [comment]
Please provide screenshots with frame graphs (ctrl+f11 twice) and also provide more details on whether you feel high input latency or are just looking at the fps counter and noticing.

**Comment by akippnn** [comment]
I also have a similar issue that's only occurring since the new renderer was released. (Sorry, forgot to take frame graphs for the earlier screnshots, but they were stable for the most part.)

I couldn't really find any reproducible results. Here's what I've done so far

Earlier, the the framerate seemed to be locked at 72FPS no matter what renderer I set it on, except for Vulkan, however, osu!lazer freezes on Vulkan the moment I press Play (same issue as #22960).
![image](https://user-images.githubusercontent.com/80670461/227794990-db6c746f-e5e8-4a57-b01c-41999666ea5b.png)

This was regardless of the game being in fullscreen mode. So I messed around to see what was causing it.

Somehow enabling G-SYNC changed how the game was behaving ever since. I set that on and now my game ran at the set in-game refresh rate regardless of whether it was in full-screen or not.
![image](https://user-images.githubusercontent.com/80670461/227795408-c938f3db-bd66-43e6-9cbe-1d0a281ac77e.png)
![image](https://user-images.githubusercontent.com/80670461/227795420-cb64de4f-389e-4a64-9349-c489403e7f87.png)

But now I set the game to the global default and restarted osu again, to find that the issue was somehow fixed but only in exclusive fullscreen though. Otherwise, the game would be stuck in 144FPS until then. Here's the frame graphs for the latter case:
![image](https://user-images.githubusercontent.com/80670461/227796469-a2ba0885-ce80-4f60-bf71-3378ca1aa9c7.png)

Can't really point finger to this being a G-SYNC issue either for now, because I just turned it off entirely and turned it back on and set G-SYNC for osu again, with the same issue (144FPS on non-exclusive fullscreen) still being the case.

Logs (nothing interesting in particular from what I've seen):
[updater.log](https://github.com/ppy/osu/files/11072646/updater.log)
[database.log](https://github.com/ppy/osu/files/11072647/database.log)
[input.log](https://github.com/ppy/osu/files/11072648/input.log)
[performance.log](https://github.com/ppy/osu/files/11072649/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11072650/runtime.log)

**Comment by longnguyen2004** [comment]
I'd say this is very much driver dependent. Try resetting all the settings and try again, who knows what those drivers are doing below osu itself...

**Comment by akippnn** [comment]
> I'd say this is very much driver dependent. Try resetting all the settings and try again, who knows what those drivers are doing below osu itself...

Yeah, that's the first conclusion I came to (resetting the settings that is), but it didn't really fix much since the game was still stuck in 72FPS regardless (even after a computer restart) before I started messing around with G-SYNC settings.

**Comment by mouzedrift** [comment]
> Please provide screenshots with frame graphs (ctrl+f11 twice) and also provide more details on whether you feel high input latency or are just looking at the fps counter and noticing.

| Fullscreen | Borderless fullscreen | Windowed |
| ----------- | ----------------------- | ------------ |
| ![osu!_AaWlr7kotm](https://user-images.githubusercontent.com/43358824/227797898-ddfbef02-ed4e-42e0-826e-906b3a99bbcb.png) | ![image](https://user-images.githubusercontent.com/43358824/227797963-0c0c6d87-99bc-482b-bf52-7074bc0d8740.png) | ![image](https://user-images.githubusercontent.com/43358824/227798023-1d5b32d6-27f7-45a8-8c06-8e47b61c17f6.png) |

The mouse input latency in fullscreen is pretty bad and very delayed (borderless and windowed feel more responsive than fullscreen), but the keyboard input latency seems fine.

edit: The keyboard input latency seems to be worse too, but I just haven't noticed it as much as the mouse input latency.

**Comment by longnguyen2004** [comment]
For Borderless and Windowed: You might be seeing DWM VSync in action, which can be disabled by passing [`DXGI_PRESENT_ALLOW_TEARING`](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/dxgi-present#:~:text=in%20Windows%C2%A08.1.-,DXGI_PRESENT_ALLOW_TEARING,-0x00000200UL) to the DXGI `Present` function.

I saw in the framework's code that there's an [`AllowTearing`](https://github.com/ppy/osu-framework/blob/745fce6716cdcd3c945231d7fd616c4a8e6b6168/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs#L38-L42) flag, but [it's only set when the game is in full screen mode](https://github.com/ppy/osu-framework/blob/745fce6716cdcd3c945231d7fd616c4a8e6b6168/osu.Framework/Platform/GameHost.cs#L484). You can use something like Special K to inject the flag (only a temporary solution, proper fix needs to be done in code). With that, VSync is properly off in borderless mode.

![image](https://user-images.githubusercontent.com/15729831/227828115-7ddb555f-d838-4517-9f94-7c24fa58e4bb.png)

As for Fullscreen mode, that's definitely some driver stuff going on, because I can't reproduce it, so I highly doubt osu! can do anything here.

**Comment by peppy** [comment]
We *intentionally* don't want to allow tearing in borderless, as that's generally the expectation of what "borderless" means to a user.

**Reply by Rubydesic** [reply]
I don't understand this. Changing the renderer (from DirectX 11 to OpenGL) uncaps the framerate in borderless mode, and the framerate is also not capped in osu stable in borderless mode. Are you suggesting that both osu stable and the OpenGL renderer have broken behavior and this is the correct one? Forcing the capped framerate introduces significant and noticeable input lag. 

The only reason that I (and I'm sure many others) use borderless mode is so that we can interact with another screen without minimizing the game, nothing to do with tearing.

**Comment by longnguyen2004** [comment]
Hmm we should probably enable that anyway, because some users might have a borked fullscreen exclusive mode (like @mouzedrift). Besides, the new flip model + waitable swapchain is supposed to remove the need for fullscreen exclusive, which is further exemplified by the fact that UWP apps can't use fullscreen exclusive at all. IMO, we should enable `AllowTearing` all the time, and users who don't want to see tearing can use VSync.

**Comment by peppy** [comment]
That's not the direction we will be taking, but thanks for your thoughts.

Avoiding tearing is paramount for a normal user experience.

**Comment by Andreyf20** [comment]
Same problem, tried to play around with gsync and got it to work (unlimited) but only in full screen, I decided to go with opengl for now which seems to be working in windowed for me.

**Comment by longnguyen2004** [comment]
Seems like people are using fullscreen windowed not to avoid tearing, but just to be able to Alt+Tab without the screen flickering. I think we can safely assume that the people using unlimited FPS already know the drawbacks and consequences of it, so we can probably allow tearing in fullscreen windowed, enable VSync by default, and give a big fat warning if people switch away from it.

Remember, if there's no tearing, then VSync must be somewhere in the chain. It might be game VSync, DWM VSync, driver VSync, OpenGL triple buffering, etc etc., but it's there, even if the in-game counter tells you otherwise, so I'd rather make it explicit that VSync is enabled, rather than having DWM silently enabling it. Some people just want to see the big numbers, even though [it's irrelevant for lazer now](https://github.com/ppy/osu/wiki/Latency-and-unlimited-frame-rates), but since we can give them the option to do that, why not?

**Comment by Layendan** [comment]
I can recreate this bug with the latest version of osu.

However, if I switch framerate from 2x to vsync, then back to 2x, the issue is solved for the session/user restarts game.
Although better, I still feel as if there is input lag on both keyboard and mouse when using this fix.

Can anyone recreate this or am I the only one where changing framerates fixes the problem?

**Comment by longnguyen2004** [comment]
> Can anyone recreate this or am I the only one where changing framerates fixes the problem?

Not reproducible. Also, this is specific to the Direct3D 11 mode, it works fine on OpenGL.

**Comment by Layendan** [comment]
> Not reproducible. Also, this is specific to the Direct3D 11 mode, it works fine on OpenGL.

I have the same problem with OpenGL as well as the Legacy render.

**Comment by peppy** [comment]
There's a high likelihood that you have a driver setting overriding things.

**Comment by Layendan** [comment]
> There's a high likelihood that you have a driver setting overriding things.

Although I would agree with that, I never changed any of my driver settings.

I tried resetting them when it was mentioned in this thread, but that didn't change anything.
I also tried reinstalling my graphics drivers and updating them, but that did not help.

I've been playing this game for 2 years with no problem, then I boot up the game and the issue starts showing up.

It might also not be osu since there wasn't an update since, but this issue seemed very similar to what was happening to me and was hoping to see if there was something I could do to help fix it.

**Comment by Kulei666** [comment]
Have the same issue, never changed any of my driver settings. It happened after an update, but I am not sure what version I was updating from. Regardless of the VSync/2x/4x/8x/unlimited setting FPS draw drops to 60/60, with ~17ms of lag input, both on keyboard and mouse. It seems to happen whenever I click play and it loads up maps (my fps drops to 1 and then it only bounces back to 60) and whenever I start a beatmap (same). Temporary fix is to pause the beatmap and alt+tab and the FPS bounces back to the set value, however if there is any stuttering during gameplay (even above the max fps limit, best visible with uncapped) the game might snap to 60 for some reason. It happens both in DirectX and OpenGL. Very rarely it uncaps itself during gameplay for a few seconds, but then usually snaps back to the mysterious 60. (Edit: 60 is also my monitor refresh rate)

**Comment by Layendan** [comment]
So I've been playing with a few apps that I have and am planning on doing a full wipe of my pc to test some more, but I've noticed that if I have a chromium based browser (might be any browser, but only tested with Chrome and Edge) open while opening the game, it will consistently cap the fps, while if the browser is closed it will only do it once (Tested only on Legacy OpenGL). Then you can alt+tab out and back in, then the fps does not cap anymore.

@Kulei666 what browser are you using and would you mind please trying this out to see if that works for you as well?

**Comment by peppy** [comment]
This sounds like you are

- Running borderless or windowed
- Have the chrome instance open on a second display with 60hz refresh rate.

This is a known issues on windows (out of our control) if so.

**Reply by Layendan** [reply]
> Running borderless or windowed
Have the chrome instance open on a second display with 60hz refresh rate.

I am running in fullscreen and I only have 1 monitor at the moment.

**Comment by Kulei666** [comment]
@Layendan @peppy Did some more testing today, turned off all apps and changed NVIDIA's control panel performance mode into Max Performance. None of that helped. Alt+tabbing still solves the issue (I am playing on fullscreen). Closely examining the fps I made an observation (on two 6-star maps) it usually caps after a beatmap break (by break i mean while music is playing but no circles), what happens during a break is that fps first drop and cap at 180, then drop and cap at 120 and when the beatmap resumes from the break it drops to 60 fps again.

**Comment by peppy** [comment]
if it's exactly 60 then it's vsync

**Comment by Kulei666** [comment]
Yes, it's exactly 60 fps, it should be 8x that, 480fps. More testing and: When it drops below 420fps it immediatelly snaps to 60fps. 3 seconds before a break end it drop to 180, 1 seconds before break end it drops to 120, 2 seconds into the map it drops to 60fps and snaps perfectly onto it until alt+tabbed. How do I fix that? I forced Vertical sync Off in the Nvidia control panel and don't have Vsync active in osu!lazer

**Comment by Kulei666** [comment]
Another weird fix - recording the display with OBS makes the problem not happen at all, doesn't snap to 60fps.
Edit: Just having OBS open fixes it.

**Reply by peppy** [reply]
OBS has a setting to force vsync off i believe.

This is up to you to figure out. Something on your system is forcing vsync on. It could be anything (recording software, overlays, nvidia profiles), but it's definitely not osu!.

**Reply by Kulei666** [reply]
Permanently fixed and fps drastically stabilised by changing "Low Latency Mode" in NVIDIA Control Panel from "Ultra" to "Off".

**Reply by Layendan** [reply]
> Permanently fixed and fps drastically stabilised by changing "Low Latency Mode" in NVIDIA Control Panel from "Ultra" to "Off".

I'll try this tomorrow when I'll have free time, see if that does anything

**Reply by Layendan** [reply]
I looked in my NVIDIA control panel and I already had it set to 'Off', but still got the cap at 144hz (monitor refresh rate)

**Reply by Kulei666** [reply]
I had a clean Windows 10 install this week after my SSD got corrupted. On clean install the fps is capping itself at 120 or 180 fps even though it's set at 8x refresh rate. I also checked control panel for Vsync, Low Latency Mode and Power Management Mode (the ones that fixed it before), but it all looks okay. Edit: Having OBS open seems to fix it again, but with some stuttering present.

**Comment by mouzedrift** [comment]
no idea if something changed on osu's side but it no longer seems to cap my fps at my monitor's refresh rate

**Reply by Kulei666** [reply]
the same update fixed it for me too

**Reply by Layendan** [reply]
Sadly I still have the problem on DirectX. Was looking promising but it came back in the middle of playing a map.

edit: It fixed the problem on OpenGl and (legacy)
